### PR TITLE
Add proof reconstruction for ProofRewriteRule::MACRO_QUANT_VAR_ELIM_EQ

### PR DESCRIPTION
**Pull request by ajreynol**
Handles nearly all cases of this rule, although 4/126 are instances of a more complex rewrite ("datatype tester expansion") which need separate treatment.

### CODE REVIEW

**Review by abdoo8080 — CHANGES_REQUESTED** [review]
Thanks for the PR! Looking back at the documentation for `MACRO_QUANT_VAR_ELIM_EQ`, I think its formula should be:
```
   *   \forall Y_1 x Y_2.\> F = \forall Y_1 Y_2.\> F \{ x \mapsto t \}
```

File: src/theory/quantifiers/quantifiers_rewriter.cpp [hunk 1]
```diff
@@ -1287,14 +1323,19 @@ bool QuantifiersRewriter::getVarElimInternal(Node body,
...
       {
         return true;
       }
     }
     return false;
   }
-  return getVarElimLit(body, n, pol, args, vars, subs);
+  if (getVarElimLit(body, n, pol, args, vars, subs))
+  {
+    lits.emplace_back(pol ? n : n.notNode());
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/quantifiers_rewriter.cpp:1335]: No, negate implicitly does double negation elimination, which complicates things.

Lits is just a way to remember what the literal was that justified the substitution, e.g.
```
vars = {x}
subs = {y-z}
lits = {x+z=y}

File: proofs/eo/cpc/rules/Quantifiers.eo [hunk 2]
```diff
@@ -139,3 +139,44 @@
...
+; - x T: The variable we are eliminating
+; - F Bool: The body of the quantified formula in question.
+; return: >
+;    The result of eliminating x from F. This method does not evaluate if this
+;    is not a variable elimination, i.e. if F does not begin with a disequality
+;    between x and a term not containing x.
+(program $mk_quant_var_elim_eq ((T Type) (x T) (t Type) (F Bool :list))
+  (T Bool) Bool
+  (
+  (($mk_quant_var_elim_eq x (not (= x t)))        ($mk_quant_var_elim_eq_subs x t false))
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply proofs/eo/cpc/rules/Quantifiers.eo:166]: Done

File: src/rewriter/basic_rewrite_rcons.cpp [hunk 3]
```diff
@@ -626,6 +633,157 @@ bool BasicRewriteRCons::ensureProofMacroQuantPartitionConnectedFv(
...
+                     << std::endl;
+  Assert(body1.getKind() == Kind::FORALL);
+  Node eqLit = vars[0].eqNode(subs[0]).notNode();
+  Node lit = lits[0].negate();
+  if (eqLit != lit)
+  {
+    Node equivLit = eqLit.eqNode(lit);
+    cdp->addTrustedStep(
+        equivLit, TrustId::MACRO_THEORY_REWRITE_RCONS_SIMPLE, {}, {});
+  }
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/rewriter/basic_rewrite_rcons.cpp:720]: No, these are two different things that have to be performed if the equality literal is different from the solved literal.

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/rewriter/basic_rewrite_rcons.cpp]: I've deleted it now.

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**